src/StdCheats.sol

```diff
@@ -228,7 +228,7 @@ abstract contract StdCheatsSafe {
         // address), but the same rationale for excluding them applies so we include those too.
 
         // These should be present on all EVM-compatible chains.
-        vm.assume(addr < address(0x1) || addr > address(0x9));
+        vm.assume((addr < address(0x1) || addr > address(0x9)) && addr != address(vm));
```

Comment:
Should we also exclude the `console.log` address here, which is `0x000000000000000000636F6e736F6c652e6c6f67`?

Though, both vm and console aren't technically precompiles in the sense meant by this function, so perhaps we should instead go with something like [this](https://twitter.com/msolomon44/status/1622596868105732096):

```solidity
enum AddressType {
  ZeroAddress,
  Precompiles
  Payable
  NonPayable
  ForgeAddresses
  ...whatever else
}

function assumeAddressIsNot(AddressType addressType) internal { ... }
function assumeAddressIsNot(AddressType addressType1, AddressType addressType2) internal { ... }
// repeat for length of the enum
```

Reply:
I like this approach much better. Will rework.

Reply:
Actually I really like the array version. Less code and easier testing. 

Getting a lot of issues testing the overloaded one. 
```
[FAIL. Reason: The `vm.assume` cheatcode rejected too many inputs (65536 allowed)]
```

Reply:
> Prefer the array version personally. Less code and easier to test

How is the UX with the array version? Main reason I though overloads would be better is:

```solidity
// This seems like good UX for devs
assumeAddressIsNot(1, 2, 3);

// This seems like bad UX (using a uint instead of enum for brevity in this example)
uint[] memory x = new uint[](3);
x[0] = 1;
x[1] = 2;
x[2] = 3;
assumeAddressIsNot(x);
```

> How do you propose to assert that an address is payable vs nonpayable and why include those in the enum exactly? Not positive if there's an optimal way to do this other than doing `call` on the address with some `msg.value` and checking `success`.

In the forge-std `assumePayable` method we just check `success` on a zero value call. But on second thought I'm not sure if our tests are sufficiently robust, e.g. if a contract has a non-payable `fallback` and no `receive`, that contract is not payable, and I think our `assumePayable` would wrongly return `true`. 

Better way is probably what you said, specifically: 
- Return true if address has no code
- Otherwise, ensure recipient's current ETH balance is less than maxUint, deal 1 wei to test contract, send it to recipient and check success, reset test contract and recipient balances to their original values.

> In both impl's, I have functions for each enum type that I check against in `assumeAddressIsNot` but both this and the overloaded version fail with too many `vm.assume` rejects when using a fuzz test. What do you think of this method of testing where instead of calling `assumeAddressIsNot` all the derivative functions are called?

If it helps, we can raise the `assume` limit for that test only using the [inline config](https://book.getfoundry.sh/reference/config/inline-test-config) approach. I think the most robust way would be something like the below, so if a new address type is added it's automatically tested

```solidity
function testAssumeAddressIsNot(address addr) external {
  for (uint8 i = 0; i < type(AddressType).max; i++) {
    assumeAddressIsNot(AddressType(i));
  }
  assertTrue(addr != address(0));
  assertTrue(
      addr < address(1) || (addr > address(9) && addr < address(0x4200000000000000000000000000000000000000))
            || addr > address(0x4200000000000000000000000000000000000800)        
    );
  assertTrue(addr != 0x7109709ECfa91a80626fF3989D68f67F5b1DD12D && addr != 0x000000000000000000636F6e736F6c652e6c6f67);

}
```

Reply:
Well `Payable` and `NonPayable` are mutually exclusive right, so we wouldn't need to iterate through the whole enum but instead n - 1?

Reply:
> That for loop is valid Solidity? Don't think there's a built-in way to get the length of an enum and it doesn't seem to work in either my forge test or chisel.

Hm it should work, see the [solidity docs](https://docs.soliditylang.org/en/latest/types.html#enums):
> Using `type(NameOfEnum).min` and `type(NameOfEnum).max` you can get the smallest and respectively largest value of the given enum.

---

> `Payable` and `NonPayable` are mutually exclusive right, so we wouldn't need to iterate through the whole enum but instead n - 1?

Ah good point. So maybe we put payable/nonpayable as indexes 0 and 1 in the enum, then we have:
- main test that iterates through indicies 2-n for the non-mutually-exclusive cases
- second test for payable
- third test for nonpayable

Wdyt of that? This way new values added to the end of the enum are automatically tested

> You still have to provide the address itself as an argument to `assumeAddressIsNot` right? That's the way `assumeNoBlacklisted`, `assumeNoPrecompiles`, and  `assumePayable` are.

Oops yes, sorry forgot to include that 😅

Reply:
So even when setting `max_fuzz_rejects` to some absurdly high number like `10_000_000` which takes several minutes to complete, a fuzz test on `assumeNoNonPayable` still fails. 

So for both `assumeNoPayable` and `assumeNonPayable` I went with modified versions of the old test for the now deprecated `assumePayable`. 

Latest impl pushed.

Reply:
Awesome sounds great, thanks! Lmk if this is ready for review or if you still have changes left—will be tied up today/tomorrow but should be able to review this thursday